### ISSUE TITLE
Filter group data sync

### ISSUE THREAD

**Issue by gauthier-delacroix** [anchore/anchore#47]
Hi !

When using Anchore docker container to scan a specific image, it is useless to sync all OS group data.

I didn't find any option to filter the group data I need to sync.

Is there such an option ? If not wouldn't it be nice to have it ?

Thanks !

**Comment by zhill** [comment]
Hi @gauthier-delacroix ,

First, thanks for using Anchore, and thanks for the feedback! How are you using the scanner right now? 

I agree that if you're a shop using a specific distro then the others aren't very helpful. How about 
 something like:

`anchore feeds sync vulnerabilities centos:7`

Generally, you shouldn't have to sync more than once a day or so, so we haven't implemented a filter/selection yet since it's fairly infrequent. Once the initial sync is done the subsequent syncs are just for updates and tend to be very fast. Also, if you do decide to scan another os image and don't remember to subscribe/sync that os feed you may get incorrect results on queries, and in a lot of cases you may not know what the OS is prior to analysis.

Would a stateful service be helpful? We also have [anchore-engine](https://github.com/anchore/anchore-engine), which uses anchore as the scanner and handles feed syncs in a different way that's basically transparent. Anchore Engine is a stateful service with more functionality like webhooks, apis, and notifications. So, that's an option as well if you want a service to watch images and have analysis via an API (it's also supported in our jenkins plugin "engine-mode").

**Comment by nurmi** [comment]
Hi @gauthier-delacroix 

There isn't an option for subscribing/unsubscribing to specific 'groups' within a 'feed' in the tool currently.  I think this would be a good enhancement, but I also wanted to describe the current behavior a bit since it may alleviate a concern about the sync taking too long when it pulls all the groups.  Currently, the first sync will take some time since it will be downloading all the group data, but subsequent syncs will only pull changes for each group that have been made since the last sync.  For example, if you run an 'anchore feeds sync' a few times one after another, you'll see that the last one is very fast (and reports 'already synced') since there are no changes to the feed group since the last successful sync.  This way, if the feed sync is run frequently enough (one per day is a good starting point), each feed sync after the first one should be fast.

Is the core reason for your enhancement request to decrease sync time in general, or are there other reasons why subscribing to individual groups will be of benefit for your use case?

Best
-Dan

**Comment by gauthier-delacroix** [comment]
Hi @zhill @nurmi ,
Thanks for these both clear answers.

It seems my use case may not be fully fitting Anchore design: I'm actually using it in a CI/CD context, where an Anchore container is started and destroyed after each execution.
This forces me to make a full resync each time. In this context, I always know what I'm using it for, this is why a filter would allow me to only sync the required group, but it still would not be as optimized as your vulnerabilities cache.

Anchore Engine may be the best alternative, but I'd have to deploy a permanent stack, which I'm not able to do for now (a Rancher platform is on the line but I'm delaying until 2.0 is out).

I'll think about another solution (an Anchore Engine instance on my CI servers or something like this) so my request may be closed, except if you think my current use case should be handled.

Best regards,
Gauthier

**Comment by zhill** [comment]
Another option is to build a new container that is anchore after a feed sync, you can just use FROM anchore/cli:latest and during build just run a feed sync. That would let you keep that data in the docker cache and the updates get small on each run.

That would be faster than even a single-os sync from scratch. That might be the best temporary solution. Are you using the jenkins plugin or something else?

**Comment by gauthier-delacroix** [comment]
@zhill This is indeed a nice and simple alternative I'm gonna try ASAP.

I'm using Gitlab CI, which will allow me to easily schedule a daily|weekly build of this "pre-synced" Anchore image.

I'll try yo give you feedback tomorrow.

Thank you !

PS: Just checked Anchore Engine I didn't know yet...got a lot of new stuff to try soon ! Thanks !

**Comment by gauthier-delacroix** [comment]
I just tried your proposal.

It didn't seem to work since I don't get the "already synced" message even if running anchore just after building the image, and it still takes some time.

My Dockerfile is as simple as that:

````
FROM anchore/cli:latest
RUN anchore feeds sync
````

Both builds are done on the same host which is NTP synced so there's no timestamp issue.

Anyway, I'll continue looking but I'll quickly give a try to Anchore Engine, which is what I really need.

Thanks !

**Comment by zhill** [comment]
You'll only the 'already synced' message if the sync was done that covers the same date range, typically this means doing a sync twice (a bit weird, i know). If you use:

`RUN anchore feeds sync && anchore feeds sync`

That should result in 'already synced', but only on the same day that you build the image. Once the date increments to the next day the system will try to sync again if you run the feed sync explicitly.

But, if you're doing a daily or perhaps hourly build then perhaps you don't need to run a feed sync before scanning an image just use the data in the built image. In that case you'd only need the dockerfile you already have.

Just some options. Let me know if you have any questions/issues with anchore-engine as well.

**Comment by nurmi** [comment]
Hi @gauthier-delacroix 

Let us know if the proposed process worked out for you (run anchore feeds sync a few times to get to ‘already synced’), or if you’ve decided to go with anchore-engine instead if that’s a better fit - if we can help further let us know!

Best
Dan

**Comment by gauthier-delacroix** [comment]
Hi @nurmi 
I still keep anchore-engine as my best option but I didn't deploy it yet.
The custom "pre-synced" anchore image still don't work.
I tried syncing twice in the same Dockerfile and it never seems synced:

````
[...]
Step 2/2 : RUN date && anchore feeds sync && anchore feeds sync
 ---> Running in 5e8613ba62ba
Tue Nov 28 14:38:46 UTC 2017
initializing feed metadata: ...
syncing data for subscribed feed (vulnerabilities) ...
	syncing group data: debian:unstable: ...
	syncing group data: ubuntu:16.04: ...
	syncing group data: centos:6: ...
	syncing group data: centos:7: ...
	syncing group data: centos:5: ...
	syncing group data: ubuntu:14.10: ...
	syncing group data: ubuntu:15.04: ...
	syncing group data: debian:9: ...
	syncing group data: debian:8: ...
	syncing group data: ubuntu:12.04: ...
	syncing group data: ubuntu:17.10: ...
	syncing group data: debian:7: ...
	syncing group data: ubuntu:16.10: ...
	syncing group data: alpine:3.3: ...
	syncing group data: alpine:3.4: ...
	syncing group data: alpine:3.5: ...
	syncing group data: alpine:3.6: ...
	syncing group data: ol:6: ...
	syncing group data: ubuntu:14.04: ...
	syncing group data: ubuntu:15.10: ...
	syncing group data: ubuntu:12.10: ...
	syncing group data: ubuntu:17.04: ...
	syncing group data: ol:7: ...
	syncing group data: ubuntu:13.04: ...
	syncing group data: ol:5: ...
skipping data sync for unsubscribed feed (packages) ...
syncing data for subscribed feed (vulnerabilities) ...
	syncing group data: debian:unstable: ...
	syncing group data: ubuntu:16.04: ...
	syncing group data: centos:6: ...
	syncing group data: centos:7: ...
	syncing group data: centos:5: ...
	syncing group data: ubuntu:14.10: ...
	syncing group data: ubuntu:15.04: ...
	syncing group data: debian:9: ...
	syncing group data: debian:8: ...
	syncing group data: ubuntu:12.04: ...
	syncing group data: ubuntu:17.10: ...
	syncing group data: debian:7: ...
	syncing group data: ubuntu:16.10: ...
	syncing group data: alpine:3.3: ...
	syncing group data: alpine:3.4: ...
	syncing group data: alpine:3.5: ...
	syncing group data: alpine:3.6: ...
	syncing group data: ol:6: ...
	syncing group data: ubuntu:14.04: ...
	syncing group data: ubuntu:15.10: ...
	syncing group data: ubuntu:12.10: ...
	syncing group data: ubuntu:17.04: ...
	syncing group data: ol:7: ...
	syncing group data: ubuntu:13.04: ...
	syncing group data: ol:5: ...
skipping data sync for unsubscribed feed (packages) ...
[...]
````
I've put the date to check if the host time was good (it was, NTP synced).

Is there another way to check the db sync status ?

Thanks,

Gauthier

**Comment by nurmi** [comment]
Thanks @gauthier-delacroix 

I know it may seem counter intuitive, but in order to see the ‘already synced’ output you,d need to run the command a third time - the first sync gets you the initial set, the second will get you any deltas for the day, and the third would do nothing (already synced).  You could do this to verify, but the third command isn’t necessary to run since it doesn’t do anything (other than to report already synced).

If you ever want to manually verify, you can look in .anchore/feeds/ and review the files there, which include a time stamp range in their file names.

Best
Dan

**Comment by gauthier-delacroix** [comment]
Thanks @nurmi !

I just tried launching an analysis after launching sync twice on my custom anchore image build and final sync on the built container has indeed been instantaneous. It saved one minute on the job duration :D

I just have to schedule a daily build of my image.

Regards,
Gauthier
